demo/Demo.cpp

```diff
@@ -34,157 +36,99 @@ void Demo::RunFrame()
+    render::Primitive* shape = new render::Box(
+        glm::translate(glm::mat4(1), glm::vec3(body.linearMotion.position)),
+        glm::vec3(0.439, 0.502, 0.565),
+        render::Box::Axes{i, j, k}
     );
+    m_primitives.emplace_back(object, shape);
+    m_pGravityForce->Bind(*object);
 
-    return m_objects.back();
+    return m_primitives.back();
 }
 
-void Demo::Remove(Object& object)
+void Demo::Remove(Primitive& primitive)
 {
-    if (object.body != nullptr)
+    if (primitive.physicalPrimitive)
     {
-        Particle& particle = object.body->p;
-        m_particleForceRegistry.Remove(particle);
-
-        //Remove contact generator
-        m_particleContactGenerators.erase(
-            std::find_if(m_particleContactGenerators.begin(), m_particleContactGenerators.end(),
-                [&object](auto& generator) -> bool
-                {
-                    return (object.body == &dynamic_cast<ShapeContactGenerator<RigidBodies>*>(generator.get())->rigidBody);
-                })
-        );
-
-        //Remove RB
-        for (auto it = m_rigidBodies.begin(); it != m_rigidBodies.end(); ++it)
-        {
-            if (&it->p == &particle)
-            {
-                m_rigidBodies.erase(it);
-                break;
-            }
-        }
-
-        //Remove Particle
-        for (auto it = m_particles.begin(); it != m_particles.end(); ++it)
-        {
-            if (&*it == &particle)
-            {
-                m_particles.erase(it);
-                break;
-            }
-        }
+        m_pGravityForce->Unbind(*primitive.physicalPrimitive);
     }
 
-    //Remove object
-    for (auto it = m_objects.begin(); it != m_objects.end(); ++it)
-    {
-        if (&*it == &object)
-        {
-            m_objects.erase(it);
-            break;
-        }
-    }
+    m_primitives.remove_if([&primitive](Primitive& p) { return &primitive == &p; });
 }
 
-Demo::Object::Object(RigidBody* body, render::primitive::Primitive* shape)
-    : body(body)
-    , shape(shape)
+Demo::Primitive::Primitive(scene::Primitive* body, render::Primitive* shape)
+    : physicalPrimitive(body)
+    , renderPrimitive(shape)
 {
 }
 
 Demo::Demo()
-    : m_pRenderer(render::Renderer::GetInstance())
-    , m_particleWorld(m_particles,
-        m_particleForceRegistry,
-        m_particleContactGenerators,
-        glm::pow2(maxParticles),
-        maxParticles)
-    , m_gravityForce(glm::dvec3{0, -9.8, 0})
+    : m_scene(scene::Scene::GetInstance())
+    , m_renderer(render::Renderer::GetInstance())
 {
+    m_scene.Initialize(scene::AssetManager::GetInstance());
+    m_pGravityForce = std::make_unique<scene::Force<force::StaticField>>(force::StaticField(glm::dvec3{ 0, -9.8, 0 }));
 }
 
 void Demo::ComputeFrame(double duration)
 {
-    m_particleWorld.StartFrame();
-    m_particleWorld.RunPhysics(duration);
+    //Compute physical data
+    m_scene.ComputeFrame(duration);
```

Comment:
`m_scene` is not actually touched anywhere, aside from this method and the constructor. And in methods like `MakeBox` and others you don't add new primitives to `m_scene`. Also, `m_pGravityForce` is not registered as a force in `m_scene`. But physics work in the demo. What am I missing here?

Reply:
You missing the fact that scene is a singleton. It has high-level API, which allows indirect creation of objects, such as Force, Body classes and etc, and direct low-level API, on which previously mentioned classes are based, which is a number of public methods: MakeBody, GetBodu etc. So scene being singleton is a private field of every SceneObject based class, and all such classes are calling its low-level API under the hood. Nothing stops you from calling it directly or making your own class that will work with the scene directly.

Reply:
Ok, got it.